aes.go

```diff
@@ -38,63 +30,160 @@ func NewAESCipher(key []byte) (cipher.Block, error) {
 	if err != nil {
 		return nil, err
 	}
-	return &aesCipher{c}, nil
+	var block cipher.Block
+	cbcSupportes := loadCipher(kind, cipherModeCBC) != nil
+	ctrSupportes := loadCipher(kind, cipherModeCTR) != nil
+	gcmSupportes := loadCipher(kind, cipherModeGCM) != nil
+	aes := aesCipher{c}
+	switch {
+	case cbcSupportes && ctrSupportes && gcmSupportes:
+		block = cipherWithCBC_CTR_GCM{aes,
+			cipherWithCBC{aes},
+			cipherWithCTR{aes},
+			cipherWithGCM{aes},
+		}
+	case cbcSupportes && ctrSupportes && !gcmSupportes:
+		block = cipherWithCBC_CTR{aes,
+			cipherWithCBC{aes},
+			cipherWithCTR{aes},
+		}
+	case cbcSupportes && !ctrSupportes && gcmSupportes:
+		block = cipherWithCBC_GCM{aes,
+			cipherWithCBC{aes},
+			cipherWithGCM{aes},
+		}
+	case !cbcSupportes && ctrSupportes && gcmSupportes:
+		block = cipherWithCTR_GCM{aes,
+			cipherWithCTR{aes},
+			cipherWithGCM{aes},
+		}
+	case cbcSupportes && !ctrSupportes && !gcmSupportes:
```

Comment:
When this kind of thing is necessary, personally I think sticking with the bool pattern is easier to read/review. Also, `supportes` -> `supported`:

```go
	var block cipher.Block
	cbcSupported := loadCipher(kind, cipherModeCBC) != nil
	ctrSupported := loadCipher(kind, cipherModeCTR) != nil
	gcmSupported := loadCipher(kind, cipherModeGCM) != nil
	aes := aesCipher{c}
	switch {
	case cbcSupported && ctrSupported && gcmSupported:
		block = cipherWithCBC_CTR_GCM{aes,
			cipherWithCBC{aes},
			cipherWithCTR{aes},
			cipherWithGCM{aes},
		}
	case cbcSupported && ctrSupported && !gcmSupported:
		block = cipherWithCBC_CTR{aes,
			cipherWithCBC{aes},
			cipherWithCTR{aes},
		}
	case cbcSupported && !ctrSupported && gcmSupported:
		block = cipherWithCBC_GCM{aes,
			cipherWithCBC{aes},
			cipherWithGCM{aes},
		}
	case cbcSupported && !ctrSupported && !gcmSupported:
		block = cipherWithCBC{aes}
	case !cbcSupported && ctrSupported && gcmSupported:
		block = cipherWithCTR_GCM{aes,
			cipherWithCTR{aes},
			cipherWithGCM{aes},
		}
	case !cbcSupported && ctrSupported && !gcmSupported:
		block = cipherWithCTR{aes}
	case !cbcSupported && !ctrSupported && gcmSupported:
		block = cipherWithGCM{aes}
	case !cbcSupported && !ctrSupported && !gcmSupported:
		block = aes
	default:
		panic("unreachable")
	}
```

(Simpler pattern: the first half should have CBC, the second half should not; the first half of each half should have CTR, the others should not; every other one should have GCM.)

My mind goes to code generation, but it would be the first in the repo (other than vectors_test.go), and this is maybe not quite bad enough to open *that* box... 😅

Reply:
> supportes -> supported:

Done

> (Simpler pattern: the first half should have CBC, the second half should not; the first half of each half should have CTR, the others should not; every other one should have GCM.)

Agree!

> My mind goes to code generation

It will be necessary if we ever add another mode, agree. For now it is manageable.